docs/lib/dynamic_config.md

```diff
@@ -0,0 +1,579 @@
+```
+
+and then the settings can be externally overridden as:
+
+```bash
+export MYAPP_KEY=value
+export MYAPP_KEY__SUBKEY=value
+export MYAPP_DATABASES__default__NAME=mydb
+export MYAPP_NUMBER=42
+export MYAPP_LIST="[1, 2, 3]"
+export MYAPP_DICT='@json {"key": "value"}'
+# or a valid toml
+export MYAPP_DICT='{key="value"}'
+export MYAPP_BOOL=true
+export MYAPP_FLOAT=3.14
+export MYAPP_INTERPOLATE="@format The key value is {this.KEY}"
+export MYAPP_INTERPOLATE_COMPLEX="@int @jinja {{ this.NUMBER * 4}}"
+export MYAPP_INSTALLED_APPS="@merge new_app_to_add"
+```
+
+More on [dynaconf] documentation.
+
+
+### Export settings back to django settings
+
+The export function will take all variables from the dynaconf object and export them back to django settings so it will be available as `django.conf.settings`:
+
+It is important to call this function at the very end of the `settings` module.
+
+```python
+export(__name__, DYNACONF)
+```
+
+### Validation
+
+Optionally, validators can be specified and checked, the recommendation is to perform the validation at the end of the `settings` module after the call to `export`.
+
+```python
+from dynaconf import Validator
+
+DYNACONF.validators.register(
+    Validator("KEY", required=True),
+    Validator("NUMBER", is_type_of=int, gte=0, lte=100),
+    Validator("LIST", len_min=1, when=Validator("OTHER", eq="value")),
+    Validator("DICT", condition=lambda x: x.get("key") == "value"),
+)
+
+validate(DYNACONF)
+```
+
+Check more on the validation section of [dynaconf] documentation.
+
+
+## Setting Settings 
+
+With Dynaconf, settings can come from a variety of sources, each service can define individually its default settings that can be either hardcoded passed directly to the factory function or loaded from a file defined as the `settings_files` option,
+then DAB after loading the defaults will conditionally inject its own required settings depending on the `INSTALLED_APPS` 
+settings, then in the settings.py each service can manipulate the dynaconf object individually in order to set new settings, load extra files, or override the defaults.
+
+DAB also provides helpers to load settings from standard locations and environment variables, and to validate the settings.
+
+### Defaults
+
+Assuming `settings_files=["defaults.py"]` is passed to the factory function, the `defaults.py` file will be loaded by default.
+
+`myapp/defaults.py`:
+```python
+DATABASES = {
+    "default": {
+        "ENGINE": "django.db.backends.sqlite3",
+        "NAME": "db.sqlite3",
+    }
+}
+DEBUG = False
+ANYTHING = "value"
+```
+
+Each service can name that file differently, but the `defaults.py` is a convention, it is also possible to pass multiple files to the `settings_files` option, so dynaconf loads them in order and applies the merging strategy to overrride the previous loaded settings, the `settings_files` option can also be paths to files in formats such as `yaml`, `json`, `toml`, `ini`, or python files.
```

Comment:
I feel like we should be more prescriptive here. i.e. we should include defaults.py by default and not let it be overridden. This will give the "platform" more coherance.

Reply:
I wish we could, but one requirement I got is to not disrupt, so I had to implement things in a way it doesn't break compatibility with existing codebase.

For example, Hub does not have a defaults.py, and will not have, because it comes from `pulpcore.app.settings`

GATEWAY names the dev file `settings_dev.py` and it must be kept, because otherwise the RPM building process will need to change to handle `development_defaults.py`.

The way to make it happen and not break or create friction is to let each service to outline how settings are organized.